File: src/main/java/com/alxan/noteefy/log/DefaultLogger.java
package com.alxan.noteefy.log;

import com.alxan.noteefy.common.Identifiable;
import com.alxan.noteefy.notification.Level;
import com.alxan.noteefy.notification.Notification;
import com.alxan.noteefy.notification.NotificationCenter;
import com.alxan.noteefy.notification.exception.NoteefyException;

import java.util.UUID;

public class DefaultLogger implements Logger {
    private UUID nodeId;

    public DefaultLogger(Identifiable node) {
        nodeId = node.getUUID();
    }

    public void log(String log, Level level) {
        if (level == Level.ERROR) {
            error(log);
            return;
        }
        Notification<String> notification = new Notification<>(log, level);
        NotificationCenter.notify(notification);
    }

    @Override
    public void error(String log) {
        NoteefyException exception = new NoteefyException(log, nodeId);
        Notification<? extends NoteefyException> notification = new Notification<>(exception, Level.ERROR);
        NotificationCenter.notify(notification);
    }

    @Override
    public void trace(String log) {
        log(log, Level.TRACE);
    }

    @Override
    public void warn(String log) {
        log(log, Level.WARN);
    }

    @Override
    public void debug(String log) {
        log(log, Level.DEBUG);
    }

    @Override
    public void info(String log) {
        log(log, Level.INFO);
    }
}

File: src/log/DefaultLogger.kt
package log

import notification.NotificationCenter
import org.apache.logging.log4j.Level

class DefaultLogger: Logger {
    override fun log(level: Level, message: String) {
        NotificationCenter.notify(message)
    }

    override fun info(message: String) = log(Level.INFO, message)
    override fun warn(message: String) = log(Level.WARN, message)
    override fun debug(message: String) = log(Level.DEBUG, message)
    override fun trace(message: String) = log(Level.TRACE, message)
    override fun error(message: String) = log(Level.ERROR, message)
}